internal/controllers/configuration/kongupstreampolicy_controller.go

```diff
@@ -0,0 +1,255 @@
+			r.Log.Error(err, "Failed to retrieve KongUpstreamPolicy in watch predicates", "KongUpstreamPolicy", fmt.Sprintf("%s/%s", service.Namespace, policyName))
+		}
+		return []reconcile.Request{}
+	}
+
+	return []reconcile.Request{
+		{
+			NamespacedName: k8stypes.NamespacedName{
+				Namespace: service.Namespace,
+				Name:      policyName,
+			},
+		},
+	}
+}
+
+// doesServiceUseUpstreamPolicy filters out all the services not referencing KongUpstreamPolicies.
+func doesServiceUseUpstreamPolicy(obj client.Object) bool {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return false
+	}
+	if service.Annotations == nil {
+		return false
+	}
+	_, ok = service.Annotations[kongv1beta1.KongUpstreamPolicyAnnotationKey]
+	return ok
+}
+
+// -----------------------------------------------------------------------------
+// KongUpstreamPolicy Controller - Reconciliation
+// -----------------------------------------------------------------------------
+
+// +kubebuilder:rbac:groups=configuration.konghq.com,resources=kongupstreampolicies,verbs=get;list;watch
+// +kubebuilder:rbac:groups=configuration.konghq.com,resources=kongupstreampolicies/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=httproutes,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch
+
+// Reconcile processes the watched objects.
+func (r *KongUpstreamPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("KongV1beta1KongUpstreamPolicy", req.NamespacedName)
+
+	// get the relevant object
+	kongUpstreamPolicy := new(kongv1beta1.KongUpstreamPolicy)
+
+	if err := r.Get(ctx, req.NamespacedName, kongUpstreamPolicy); err != nil {
+		if apierrors.IsNotFound(err) {
+			kongUpstreamPolicy.Namespace = req.Namespace
+			kongUpstreamPolicy.Name = req.Name
+
+			return ctrl.Result{}, r.DataplaneClient.DeleteObject(kongUpstreamPolicy)
+		}
+		return ctrl.Result{}, err
+	}
+	log.V(util.DebugLevel).Info("Reconciling resource", "namespace", req.Namespace, "name", req.Name)
+
+	// clean the object up if it's being deleted
+	if !kongUpstreamPolicy.DeletionTimestamp.IsZero() && time.Now().After(kongUpstreamPolicy.DeletionTimestamp.Time) {
+		log.V(util.DebugLevel).Info("Resource is being deleted, its configuration will be removed", "type", "KongUpstreamPolicy", "namespace", req.Namespace, "name", req.Name)
+
+		objectExistsInCache, err := r.DataplaneClient.ObjectExists(kongUpstreamPolicy)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if objectExistsInCache {
+			if err := r.DataplaneClient.DeleteObject(kongUpstreamPolicy); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{Requeue: true}, nil // wait until the object is no longer present in the cache
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// get all the services that reference this UpstreamPolicy
+	services := &corev1.ServiceList{}
+	err := r.List(ctx, services,
```

Comment:
Would this be a high cost operation to list all services in a namespace? If so ,we can consider caching the reference relations. (Maybe in other issue)

Reply:
yep, we are already listing only the services that reference the `UpstreamPolicy` under reconciliation.